STYLE-GUIDE.md

```diff
@@ -0,0 +1,285 @@
+
+### Prefer `map` to `fmap` or `<$>`
+
+When operating on a list, using `map` tells the reader "I'm transforming a list", where `fmap` tells them "I'm transforming *some* functor, and it doesn't matter which one.
+
+```haskell
+appendFoo t = t <> "-foo"
+txts = ["Hello", "darkness", "my", "old", "friend"]
+
+-- Ambiguous
+appendFoo <$> txts
+-- Slightly better
+fmap appendFoo txts
+-- Clear and Obvious
+map appendFoo txts
+```
+
+### Don't use `&` or `<&>`
+
+`&` and `<&>` are the flipped versions of `$` and `<$>`, and it is tempting to use them, since they read from left-to-right, like english.  However, the usefulness of
+`$` is to remove parentheses, and `&` makes that a lot harder:
+
+```haskell
+-- Raw parens
+foo (bar baz (quux))
+-- Using $
+foo $ bar baz $ quux
+-- Using &
+quux & bar baz & foo
+```
+
+`<$>` is meant to mimic `$`, but for functors.  Therefore, if we use `$` instead of `&`, we should also use `<$>` instead of `<&>`.
+
+## Data Types
+
+### Prefer `newtype`s to `type`s
+
+`newtype` values are zero-cost to construct and eliminate, and provide more informative error messages than `type` synonyms. Only use `type` for convenience aliases to existing types.
+
+### Don't use `String`
+
+`String` is almost always the wrong choice. If your type represents human-readable strings, use `Text`; if you have a blob of bytes, use `ByteString`. `-XOverloadedStrings` is enabled globally to make this easy.
+
+### Only use record selectors on single-constructor types
+
+The following code generates two partial functions, which is bad:
+
+``` haskell
+data Bad = Evil { getInt :: Int }
+         | Bad  { getFloat :: Float }
+
+-- getFloat (Evil 1) throws an error
+-- so does getInt (Bad 1)
+```
+
+Often, this is incorrect deisgn.  You usually want something like the following three examples instead:
+
+```haskell
+-- Alternative 1
+data Good
+  = Okay Int
+  | NotBad Float
+
+-- Alternative 2
+data Alright = Alright
+  { getInt :: Maybe Int
+  , getFloat :: Maybe Float
+  }
+
+-- Alternative 3
+-- This is a little bit heavyweight, but is really strongly typed.
+newtype SpecialInt = SpecialInt { getInt :: Int }
+newtype SpecialFloat = SpecialFloat { getFloat :: Float }
+data Great
+  = NiceInt SpecialInt
+  | CoolFloat SpecialFloat
+```
+
+## Imports
+
+### Don't qualify imports with single letters
+
+Qualified imports deserve intelligent names, like any other variable.  Some common instances are caught by `hlint`:
+
+```haskell
+import Data.Text qualified as T -- bad; hlint will catch this specific case
+import Data.Text qualified as Text -- Good!
+```
+
+### Don't use implicit blanket imports
+
+Import items explicitly:
+
+```haskell
+import Control.Effect.Diagnostics -- bad, which items do you need from this module?
+import Control.Effect.Diagnostics (fatal, context)  -- Good
+```
+
+There are notable exceptions, but they are few and far between.  The `Path` module is a common exception, but there's no good justification for this.
```

Comment:
> The `Path` module is a common exception, but there's no good justification for this.

Does this mean "`Path` is a good exception" or "`Path` is a bad exception and we haven't gotten around to fixing it?". This sentence could read either way.

Reply:
It means "It's an exception we make regularly, but we don't have a good reason to make it an exception."  I'm wide open to rephrasing suggestions here, but that seems pretty clear to me.  Not sure how I would improve it.

Reply:
If there isn't a great reason to make an exception I don't think we should. Is it overkill to maintain a list of the notable exceptions? I think that would help clear up ambiguity and I don't think there are too many imports that we would have to list.

Reply:
Yeah, let's remove this note.  I'd like to check this statically, but I'm not sure if it's a GHC warning or hlint rule, or neither.

Reply:
ghc's `-Wmissing-import-lists` will do this

Reply:
Why _do_ we make `Path` an exception? Let's just fix it up. Is there a technical reason why that's difficult, or is it just something we haven't done yet?

Reply:
Just something we haven't done. Someone needs to submit a PR that enables `-Wmissing-import-lists`, and do the manual work of adding explicit imports to every module (through the code action in haskell-language-server, probably)